### PR TITLE
7657  G4P Forbedringer av kode og testdekning

### PR DESCRIPTION
**Pull request by ragnarwestad**
Disse to commitene forbedrer kodekvaliteten og testdekningen for bugfixen som gjør "Legg til periode"-knappen synlig for delt grunnlag selv med pliktige bestemmelser.

Commit 1: Refactor - Erstatt any-typer med spesifikke TypeScript-typer

  Hva: Fjerner 16+ any-typer og erstatter med eksplisitte typer.
  Hvorfor: Bedre type-sikkerhet, færre potensielle runtime-feil, og bedre IntelliSense-støtte.

  Endringer:
  - getChangedDependencies: Fullstendig typet med Record<string, unknown> og callback-signatur
  - previousFormState: Bruker ReturnType<typeof mapFormState> for automatisk typeinferens
  - behandlingID: Spesifisert som number i stedet for any
  - Fjernet unødvendige control as any casts i useFieldArray
  - Refs typet korrekt (ReturnType<typeof Utils._debounce> og Record<string, unknown>)
  - Eksplisitte typer i map-funksjoner og sorteringsfunksjoner

Commit 2: Test - Legg til tester for "Legg til periode"-knappen med delt grunnlag

  Hva: Omfattende testdekning for bugfixen og relatert funksjonalitet.
  Hvorfor: Sikrer at bugfixen fungerer som forventet og forhindrer regresjoner.

  Nye tester:

  Unit-tester (16 totalt, alle kjører ✅)

  1. medlemskapsperiodeSkjema.test.tsx (NY - 7 tester)
    - ✅ Knappen vises IKKE med pliktig bestemmelse + ikke delt grunnlag
    - ✅ Knappen vises med pliktig bestemmelse + delt grunnlag (BUGFIX)
    - ✅ Knappen vises alltid med ikke-pliktig bestemmelse
    - ✅ Kun siste periode viser knappen
    - ✅ Sannhetstabell for alle kombinasjoner
  2. aarsavregningUtenEllerDeltGrunnlag.test.tsx (UTVIDET - +4 tester)
    - ✅ erDeltGrunnlag beregning med alle kombinasjoner
    - ✅ Edge cases (undefined, null)

  E2E-tester

  3. melosys-7612-aarsavregning-delt-grunnlag.spec.ts (UTVIDET - +2 tester)
    - ✅ AC2.5: Verifiserer knapp-synlighet med § 2-7 (pliktig bestemmelse)
    - ✅ AC2.5: Verifiserer at man kan legge til periode i dette scenariet

  Testdekning:
  - ✅ Knapp-visningslogikk for alle scenarioer
  - ✅ erDeltGrunnlag-beregning
  - ✅ End-to-end flow med pliktig bestemmelse + delt grunnlag

### CODE REVIEW

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx [hunk 1]
```diff
@@ -459,7 +460,10 @@ export function AarsavregningUtenEllerDeltGrunnlagForm({
       trygdedekning,
     }));
 
-    const sorterEtterFomDato = (a: any, b: any) => {
+    const sorterEtterFomDato = (
+      a: Pick<Medlemskapsperiode, "fomDato" | "tomDato" | "trygdedekning">,
+      b: Pick<Medlemskapsperiode, "fomDato" | "tomDato" | "trygdedekning">,
+    ) => {
```
Comment (IsaSalemHame) [inline src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx:466]: Pick? Aldri hørt om, virka veldig AI'ette.. hvorfor bruker man ikke bare Medlemskapsperiode? Er det ikke i realiteten bare det?

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx [hunk 2]
```diff
@@ -53,7 +53,7 @@ const kanPeriodeSlettes = (gjeldendePeriode: Medlemskapsperiode, allePerioderILi
 export interface PeriodeElementerProps {
   redigerbart: boolean;
   control: Control;
-  field: FieldArrayWithId<FieldArrayProps, "medlemskapsperioder">;
+  field: { id: string | number; [key: string]: unknown };
```
Comment (IsaSalemHame) [inline src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx]: vet ikke om dette var så mye bedre enn før..

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx [hunk 3]
```diff
@@ -62,7 +62,7 @@ export interface PeriodeElementerProps {
   maxDate?: Date;
   minDate?: Date;
   trygdedekninger?: string[];
-  setValue: (name: string, value: any, options?: any) => void;
+  setValue: (name: string, value: unknown, options?: { shouldValidate?: boolean; shouldDirty?: boolean }) => void;
```
Comment (IsaSalemHame) [inline src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx:65]: Hva er fordel med "unknown" enn "any"? er hipp som happ spør du meg :O

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx [hunk 4]
```diff
@@ -60,15 +60,18 @@ const consoleLog = (message: string, ...args: unknown[]) => {
 
 // Hjelpefunksjon for å logge og returnere endrede avhengigheter
 const getChangedDependencies = (
-  currentDeps: Record<string, any>,
-  previousDepsRef: React.MutableRefObject<any>,
-  consoleLogCallback: any,
+  currentDeps: Record<string, unknown>,
+  previousDepsRef: React.MutableRefObject<Record<string, unknown> | null>,
+  consoleLogCallback: (message: string, ...args: unknown[]) => void,
```
Comment (IsaSalemHame) [inline src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx:65]: like mystisk som før egentlig

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx (hunk 1 above)
Comment (ragnarwestad) [reply src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx:466]: Jeg synes den er grei:

  1. Type-sikkerhet: Hvis Medlemskapsperiode endres, følger typen automatisk med
  2. Konsistens: Garanterer at typen matcher kildetypen
  3. DRY-prinsippet: Ingen duplisering av type-definisjoner

Og det er standard TypeScript, ikke noe AI :-)

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx (hunk 3 above)
Comment (ragnarwestad) [reply src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx:65]: I praksis for denne spesifikke parameteren gjør det liten forskjell, annet enn at unknown skal gi verktøy noe mer mulighet for autocomplete på feltnavnet. 

Men hvis react-hook-form hadde brukt full TypeScript-magi, ville signaturen kunne sett slik ut:

  setValue<K extends keyof FormValues>(name: K, value: FormValues[K]) => void;

Min innstilling til det er at vi har ESLint regel som gir warning på any, det er faktisk standard at det er error men jeg har slakket på den siden byggene feiler hvis ESLint feiler.

Og i de fleste tilfeller så vil det jo være mulig å bytte de med konkrete typer.

Men siden vi får *veldig* mange slike warnings (250 "Unexpected any" advarsler pr dato) er det veldig greit å ta de som bare kan settes til unknown også, sånn at antall warnings kryper nedover og de som kan rettes til noe "fornuftig" blir mer synlige.

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx (hunk 2 above)
Comment (ragnarwestad) [reply src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/komponenter/medlemskapsperiodeSkjema.tsx]: Det *ser* ikke bedre ut nei. Men slik det var blir det feil det ved input av "field={field}" i form'en.

Det blir litt følgefeil bakover:

1. Yup-skjemaene er ikke fullt typet - De bruker any[] i stedet for konkrete typer (Medlemskapsperiode[], etc.)
2. useForm mangler type-parameter - Uten useForm<Type>() kan ikke TypeScript inferere riktig type for fields fra useFieldArray
3. Det er pragmatisk - Komponenten bruker kun field.id, så vi trenger ikke full type-sikkerhet på alle feltene, og det gir komponenten tilgang til field.id (som er alt komponenten trenger)

File: src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx (hunk 4 above)
Comment (ragnarwestad) [reply src/sider/aarsavregning/stegKomponenter/vurderingAarsavregning/aarsavregningUtenEllerDeltGrunnlag/aarsavregningUtenEllerDeltGrunnlagForm.tsx:65]: Litt mer bruk av konkrete typer her da, i alle fall.
Og bruk av "unknown" av samme grunn som forklart over.
